extractors-canopycover/drone_canopycover.py

```diff
@@ -0,0 +1,717 @@
+                        poly.AddGeometry(ring)
+
+                        ref_sys = osr.SpatialReference()
+                        if ref_sys.ImportFromEPSG(int(epsg)) != ogr.OGRERR_NONE:
+                            logging.error("Failed to import EPSG %s for image file %s",
+                                          str(epsg), onefile)
+                        else:
+                            poly.AssignSpatialReference(ref_sys)
+
+                        imagefiles[onefile] = {'bounds' : poly}
+                    else:
+                        self.log_info("Image file is not georeferenced and is being skipped: " + onefile)
+
+        # Return what we've found
+        return imagefiles
+
+    # Make a best effort to get a dataset ID
+    # pylint: disable=no-self-use
+    def get_dataset_id(self, host, key, resource, dataset_name=None):
+        """Makes a best effort attempt to get a dataset ID
+
+        Args:
+            host(str): the URI of the host making the connection
+            secret_key(str): used with the host API
+            resource(dict): dictionary containing the resources associated with the request
+            dataset_name(str): optional parameter containing the dataset name of interest
+
+        Return:
+            The found dataset ID or None
+
+        Note:
+            The resource parameter is investigated first for a dataset ID. Note that if found,
+            this dataset ID may not represent the dataset_name (if specified).
+
+            If the resource parameter is not specified, or doesn't have the expected elements
+            then a dataset lookup is performed
+        """
+        # First check to see if the ID is provided
+        if resource and 'type' in resource:
+            if resource['type'] == 'dataset':
+                return resource['id']
+            elif resource['type'] == 'file':
+                if ('parent' in resource) and ('id' in resource['parent']):
+                    return resource['parent']['id']
+
+        # Look through all the datasets we can retrieve to find the ID
+        if dataset_name:
+            url = '%s/api/datasets/sorted' % (host)
+            params = {"key" : key, "limit" : 50000}
+            headers = {'content-type': 'application/json'}
+
+            response = requests.get(url, headers=headers, params=params, verify=False)
+            response.raise_for_status()
+            datasets = response.json()
+
+            for one_ds in datasets:
+                if 'name' in one_ds and 'id' in one_ds:
+                    if one_ds['name'] == dataset_name:
+                        return one_ds['id']
+
+        return None
+
+    def write_csv_file(self, resource, filename, header, data):
+        """Attempts to write out the data to the specified file. Will write the
+           header information if it's the first call to write to the file.
+
+           If the file is not available, it waits as configured until it becomes
```

Comment:
Why would the file not be available? Is it on a network file system? Or is another process in charge of creating the file?

Reply:
It may not be available because a different process has an exclusive open on it. For example, there may be 5 extractors calculating greenness and writing to the same CSV file. This is a back-off-and-wait approach to resolving the conflict.